### PR TITLE
cpuinfo: init at 0-unstable-2024-06-02

### PR DESCRIPTION
**Pull request by pawelchcki**
## Description of changes

Add libcpuinfo library and associated tools from  https://github.com/pytorch/cpuinfo.

```
cpu-info - gives information about available packages, cores, and logical processors
cache-info - gives detail information about available CPU cache
isa-info - lists available instructions sets
cpuid-dump - dumps raw cpuid information
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/cp/cpuinfo/package.nix [hunk 1]
```diff
@@ -0,0 +1,54 @@
+{
+  lib,
+  buildBazelPackage,
```
Comment (jackyliu16) [inline pkgs/by-name/cp/cpuinfo/package.nix:3]: remove this, it seems haven't been use in your file.

File: pkgs/by-name/cp/cpuinfo/package.nix [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  ];
+
+  cmakeFlags = [
+    "-DUSE_SYSTEM_GOOGLETEST=1"
+    "-DUSE_SYSTEM_GOOGLEBENCHMARK=1"
+  ];
+
+  doCheck = true;
+
+  meta = with lib; {
```
Comment (jackyliu16) [inline pkgs/by-name/cp/cpuinfo/package.nix:45]: Currnetly, use `with lib` in meta is not recommanded.

**Review by jackyliu16 — APPROVED** [review]

**Comment by jackyliu16** [comment]
Result of `nixpkgs-review pr 323890` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cpuinfo</li>
  </ul>
</details>

File: pkgs/by-name/cp/cpuinfo/package.nix (hunk 2 above)
Comment (pawelchcki) [reply pkgs/by-name/cp/cpuinfo/package.nix]: Done - thanks!

File: pkgs/by-name/cp/cpuinfo/package.nix (hunk 1 above)
Comment (pawelchcki) [reply pkgs/by-name/cp/cpuinfo/package.nix]: Thanks - sorry I got so used to linter checking my leftovers :sweat_smile: 

Removed. Thanks!

**Comment by pawelchcki** [comment]
thanks for quick review @jackyliu16 :bow: I pushed all requested changes

**Comment by jkachmar** [comment]
oh hey nice, when this merges i should hopefully be able to rebase https://github.com/NixOS/nixpkgs/pull/319880 & drop the vendored `cpuinfo` there

File: pkgs/by-name/cp/cpuinfo/package.nix [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  ];
+
+  buildInputs = [
+    gtest
+    gbenchmark
+  ];
+
+  cmakeFlags = [
+    "-DUSE_SYSTEM_GOOGLETEST=1"
+    "-DUSE_SYSTEM_GOOGLEBENCHMARK=1"
```
Comment (natsukium) [inline pkgs/by-name/cp/cpuinfo/package.nix:39]: How about setting `USE_SYSTEM_LIBS=1` instead?
https://github.com/pytorch/cpuinfo/blob/05332fd802d9109a2a151ec32154b107c1e5caf9/CMakeLists.txt#L26-L28

File: pkgs/by-name/cp/cpuinfo/package.nix [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  ];
```
Comment (natsukium) [inline pkgs/by-name/cp/cpuinfo/package.nix:40]:
```suggestion
    (lib.cmakeBool "CPUINFO_BUILD_BENCHMARKS" false)
  ];
```
I don't think we need to execute a benchmark when we package.
https://github.com/pytorch/cpuinfo/blob/05332fd802d9109a2a151ec32154b107c1e5caf9/CMakeLists.txt#L24

File: pkgs/by-name/cp/cpuinfo/package.nix [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+  passthru.tests.pkg-config = testers.testMetaPkgConfig finalAttrs.finalPackage;
+
+  nativeBuildInputs = [
+    cmake
+    pkg-config
+  ];
+
+  buildInputs = [
+    gtest
```
Comment (natsukium) [inline pkgs/by-name/cp/cpuinfo/package.nix:33]: `gtest` should go to `checkInputs`

File: pkgs/by-name/cp/cpuinfo/package.nix [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    gbenchmark
+  ];
+
+  cmakeFlags = [
```
Comment (natsukium) [inline pkgs/by-name/cp/cpuinfo/package.nix:37]:
```suggestion
  cmakeFlags = [
    (lib.cmakeBool "CPUINFO_BUILD_UNIT_TESTS" finalAttrs.finalPackage.doCheck)
    (lib.cmakeBool "CPUINFO_BUILD_MOCK_TESTS" finalAttrs.finalPackage.doCheck)
```
I recommend that we make the check phase skippable.
https://github.com/pytorch/cpuinfo/blob/05332fd802d9109a2a151ec32154b107c1e5caf9/CMakeLists.txt#L22-L23

**Comment by pawelchcki** [comment]
Thanks for very indepth review @natsukium :bow:. I've disabled the benchmark builds and used `cmakeBool` for everything as you suggested.

Also I reformatted the source using nixfmt-rfc.

**Review by mweinelt — APPROVED** [review]

File: pkgs/by-name/cp/cpuinfo/package.nix [hunk 7]
```diff
@@ -0,0 +1,52 @@
+{
+  lib,
+  fetchFromGitHub,
+  cmake,
+  pkg-config,
+  stdenv,
+  gtest,
+  gbenchmark,
```
Comment (mweinelt) [inline pkgs/by-name/cp/cpuinfo/package.nix:8]: unused

File: pkgs/by-name/cp/cpuinfo/package.nix [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+  checkInputs = [ gtest ];
+
+  cmakeFlags = [
+    (lib.cmakeBool "CPUINFO_BUILD_UNIT_TESTS" finalAttrs.finalPackage.doCheck)
+    (lib.cmakeBool "CPUINFO_BUILD_MOCK_TESTS" finalAttrs.finalPackage.doCheck)
+    (lib.cmakeBool "CPUINFO_BUILD_BENCHMARKS" false)
+    (lib.cmakeBool "USE_SYSTEM_LIBS" true)
+  ];
+
+  doCheck = true;
```
Comment (mweinelt) [inline pkgs/by-name/cp/cpuinfo/package.nix:41]:
```suggestion
  doCheck = !(stdenv.isLinux && stdenv.isAarch64);
```
For some reason the tests fail on aarch64-linux, but the programs all work.

**Review by mweinelt — DISMISSED** [review]

File: pkgs/by-name/cp/cpuinfo/package.nix (hunk 7 above)
Comment (mweinelt) [inline pkgs/by-name/cp/cpuinfo/package.nix:8]:
```suggestion
```

**Comment by mweinelt** [comment]
Please squash these commits into the package init

**Comment by pawelchcki** [comment]
Thank you all! And thanks @mweinelt for the final review ! :bow:

**Pull request merged**